libuavcan/include/libuavcan/media/interfaces.hpp

```diff
@@ -0,0 +1,255 @@
+     *                                      owned by this manager instance. This memory shall remain valid while
+     *                                      the manager object is valid. It is undefined behavior to destroy a manager
+     *                                      with open interfaces still allocated.
+     *                                      Implementations must define the semantics for calling this method multiple
+     *                                      times with the same interface_index and for calling closeInterface on
+     *                                      a shared pointer to an interface.
+     * @return libuavcan::Result::success if the interface was successfully opened and returned,
+     */
+    virtual libuavcan::Result openInterface(std::uint_fast8_t                                interface_index,
+                                            const typename InterfaceType::FrameType::Filter* filter_config,
+                                            std::size_t                                      filter_config_length,
+                                            InterfacePtrT&                                   out_interface) = 0;
+
+    /**
+     * Block for a specified amount of time or until an interface becomes ready to read or write.
+     *
+     * Note that it is allowed to return from this method even if no requested events actually happened, or if
+     * there are events that were not requested by the library.
+     *
+     * @param [in]     inout_interfaces   The interfaces to wait on.
+     * @param [in]     interfaces_length  The number of interfaces in the @p interfaces array. Note that the
+     *                                    interfaces array is not sparse. Any null pointers found before reaching
+     *                                    @p interfaces_length will cause libuavcan::Result::bad_argument to be
+     *                                    returned.
+     * @param [in]     timeout            The amount of time to wait for an event.
+     * @param [in]     ignore_write_available If true then this method will not return if interfaces become available
+     *                                    only for write.
+     *
+     * @return  libuavcan::Result::success_timeout if no events ocurred but the select operation timedout.
+     *          libuavcan::Result::success if one or more of the provided interfaces are ready for read, and if
+     *          @p ignore_write_available is false, or write.
+     *          libuavcan::Result::success_partial if one or more errors were reported for one or more interfaces.
+     */
+    virtual libuavcan::Result select(const InterfaceType* const (&interfaces)[MaxSelectInterfacesVal],
+                                     std::size_t                    interfaces_length,
+                                     libuavcan::duration::Monotonic timeout,
+                                     bool                           ignore_write_available) = 0;
```

Comment:
Do you have an actual use case in mind where `interfaces_length != MaxSelectInterfacesVal` or `MaxSelectInterfacesVal != getHardwareInterfaceCount()`? My understanding is that the abstraction can be simplified substantially without sacrificing its generality by enforcing `MaxSelectInterfacesVal == interfaces_length == getHardwareInterfaceCount()`.

Reply:
Okay.

Reply:
> My understanding is that the abstraction can be simplified substantially

I don't think removing this is going to simplify it much but I agree it is a simplification we can take.

Reply:
Actually, no. the use case appeared as soon as I removed it: dynamic selection of interfaces. If I don't do this then you have to re-compile to change how many interfaces to use in a redundant group. With this parameter this can be discovered/configured using the same binary. That seems like a reasonable feature to support given that the delta (at runtime) is about two more comparisons per interface, per invocation.

Even in safety critical systems this is important since a critical node may recover from a reset only to discover that one of the interfaces is not longer available. We would still want that node to come up with the one working interface.

Reply:
Okay, that seems valid. I see that the new interface does no longer require `MaxSelectInterfacesVal` which is good.

Reply:
Yeah. Your suggestions have yielded a much nicer abstraction IMHO. Thanks.